test/UnitWSDClient.hpp

```diff
@@ -0,0 +1,118 @@
+    UnitWebSocket(const std::shared_ptr<SocketPoll>& socketPoll, const std::string& documentURL)
+    {
+        Poco::URI uri(helpers::getTestServerURI());
+        _httpSocket = helpers::connectLOKit(socketPoll, uri, documentURL, "UnitWebSocket ");
+    }
+
+    /// Destroy the WS.
+    /// Here, we can't do IO as we don't own the socket (SocketPoll does).
+    /// In fact, we can't destroy it (it's referenced by SocketPoll).
+    /// Instead, we can only flag for shutting down.
+    ~UnitWebSocket() { _httpSocket->asyncShutdown(); }
+
+    const std::shared_ptr<http::WebSocketSession>& getWebSocket() { return _httpSocket; }
+};
+
+/// A WSD unit-test base class with support
+/// to manage client connections.
+class UnitWSDClient : public UnitWSD
+{
+public:
+    UnitWSDClient(const std::string& name)
+        : UnitWSD(name)
+    {
+    }
+
+protected:
+    const std::string& getWopiSrc() const { return _wopiSrc; }
+
+    const std::unique_ptr<UnitWebSocket>& getWs() const { return _wsList.at(0); }
+
+    const std::unique_ptr<UnitWebSocket>& getWsAt(int index) { return _wsList.at(index); }
+
+    void deleteSocketAt(int index)
+    {
+        // Don't remove from the container, because the
+        // indexes are how the test refers to them.
+        std::unique_ptr<UnitWebSocket>& socket = _wsList.at(index);
+        socket.reset();
+    }
+
+    void initWebsocket(const std::string& wopiName)
+    {
+        Poco::URI wopiURL(helpers::getTestServerURI() + wopiName + "&testname=" + getTestname());
+
+        _wopiSrc.clear();
+        Poco::URI::encode(wopiURL.toString(), ":/?", _wopiSrc);
+
+        // This is just a client connection that is used from the tests.
+        LOG_TST("Connecting test client to COOL (#" << (_wsList.size() + 1)
+                                                    << " connection): /cool/" << _wopiSrc << "/ws");
+
+        // Insert at the front.
+        const auto& _ws = _wsList.emplace(
+            _wsList.begin(),
+            Util::make_unique<UnitWebSocket>(socketPoll(), "/cool/" + _wopiSrc + "/ws"));
+
+        assert((*_ws).get());
+    }
+
+    void addWebSocket()
+    {
+        // This is just a client connection that is used from the tests.
+        LOG_TST("Connecting test client to COOL (#" << (_wsList.size() + 1)
+                                                    << " connection): /cool/" << _wopiSrc << "/ws");
+
+        // Insert at the back.
+        const auto& _ws = _wsList.emplace(
+            _wsList.end(),
+            Util::make_unique<UnitWebSocket>(socketPoll(), "/cool/" + _wopiSrc + "/ws"));
+
+        assert((*_ws).get());
+    }
+
+private:
+    /// The WOPISrc URL.
+    std::string _wopiSrc;
+
+    /// Websockets to communicate.
+    std::vector<std::unique_ptr<UnitWebSocket>> _wsList;
+};
+
+/// Send a command message to WSD from a WopiTestServer on the given connection.
+#define WSD_CMD_BY_CONNECTION_INDEX(INDEX, MSG)                                                    \
```

Comment:
Nit: does this have to be a macro? I don't immediately see something in the body that would be impossible to do with an inline function.

Reply:
Good point. This macro is used several times and the idea was to hide the internal details. My hope is to replace it with something more sensible at some point.